### PR TITLE
test(commands): edge-case hardening, parity audit, and launch checklist (Universal Commands phase 6)

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary

Phase 6 of the Universal Commands epic: adversarial verification + launch prep. Phases 0–4 are merged; this PR assumes the feature has bugs and tries to prove it, then delivers the launch artifacts.

**Headline finding: no source bugs.** Every probed degradation path in the merged implementation already degrades per spec — the 61 new tests now pin that contract so it can't silently regress.

## 1. Edge-case test sweep (new test files only — no source changes)

| File | Covers |
|---|---|
| `apps/web/src/lib/ai/core/__tests__/command-resolver-edge-cases.test.ts` | Lifecycle transitions sequenced as before/after pairs: entry page trashed AFTER creation (inject → page_trashed skip + §7.2 notice), hard-delete cascade + race window, drive membership revoked between creation and use, cross-drive entry-page access revoked (no_access, proven no content leak), disabled mid-conversation → re-enabled works again, rename (stale chip label by design). Forged tokens: `../../etc`, `<script>`, 64k label (literal, no I/O), `builtin:nonexistent`, `builtin:HELP` case-sensitivity, hostile labels never echoed into the system prompt. Chip + @mentions coexisting. |
| `apps/web/src/app/api/commands/resolve/__tests__/route-edge-cases.test.ts` | Chip resolution endpoint: hostile/64k ids → `deleted` with 200s, never reach the DB, never 500; mixed hostile+valid batches; rename returns current state without error; hard-delete race window (`entryPage: null`); drive deletion → `deleted`; membership loss → `restricted` with zero metadata leak; disabled → `ok` + `enabled:false`; id dedup. |
| `apps/web/src/app/api/commands/__tests__/route-list-edge-cases.test.ts` | Settings list (GET /api/commands): `entryPageAvailable:false` (the 'Entry page unavailable' badge data) for trashed/hard-deleted/access-revoked entry pages; title suppression (no leak); healthy siblings unaffected; no 500s. |
| `apps/web/src/lib/channels/__tests__/agent-mention-responder-edge-cases.test.ts` | **Integration with the REAL resolver** (only DB/permissions mocked, unlike the existing responder test which mocks the resolver): chip + @mention pipelines run side by side without corrupting each other; every command degradation — including a resolver DB failure — never blocks the agent response; disabled → re-enabled round trip through the channel path. |
| `apps/web/src/lib/commands/__tests__/command-chip-model-edge-cases.test.ts` | Transcript rendering under hostile input: 64k labels stay literal, forged ids degrade muted/non-navigable (never produce an attacker-controlled href), token ranges stay disjoint, §6 inert heuristic is backtracking-safe. |
| `packages/db/src/schema/__tests__/commands-schema-edge-cases.test.ts` | The cascade contract itself: entry-page/drive/user FK `onDelete: cascade`, author `set null`, scope check constraint, per-scope trigger uniqueness — if someone drops a cascade, the "chips degrade instead of erroring" guarantee breaks loudly. |

## 2–4. Launch checklist doc

`docs/specs/universal-commands-launch-checklist.md`:

- **Desktop/mobile parity audit** — 12 code-verified items with file:line references (soft-keyboard Enter-stays-newline incl. while the picker is open, `enterSelects` wiring, inputType-gated trigger detection, IME guard, visual-viewport clamping, single-tap selection, mobile sheet form, no Electron `/` accelerators, shared composer inventory, a11y wiring) + an 18-step manual QA script (step/expectation) for iOS/Android/iPad-external-keyboard/Electron.
- **Rollout plan** — the one-predicate `canUseCommands` widening, the complete (grep-verified) consumer list that lights up, 7 concrete dogfooding exit criteria keyed to real log markers (`Command resolution failed; proceeding without injection`, `[COMMANDS_*]`) and SQL checks, flip-day checklist, one-line rollback. Notes that the tracking docs' `canSeeCommandSettings` was never created (settings use `canUseCommands`) and that the gate is visibility-only, not a security boundary (server routes already enforce real permissions for everyone — accepted per spec §0).
- **Changelog draft** — the repo's only user-facing changelog surface is the marketing blog `data.ts` (publish-time-only; the old `docs/changelog` system was removed in #1044), so the draft lives in the checklist doc, clearly marked DO-NOT-PUBLISH-until-gate-widening: a ready-to-paste `BlogPost` object + short in-product copy.

## ⚠️ Phase 5 overlap note (for the orchestrator)

**Update: #1606 merged first; this branch is rebased on it (tests pass unchanged).** Phase 5 owned `packages/lib/src/commands/`, `command-resolver.ts`/`command-processor.ts` source, and the suggest route. This PR touches **none of those sources** — zero source changes anywhere; it adds new test files that import them, plus docs. Merge in either order with no conflicts. If Phase 5 changes resolver behavior (e.g. built-ins gain entry pages), these tests pin the degradation contract and will fail loudly on regression — that's the intended tripwire, not a conflict. Two new tests worth flagging to the Phase 5 agent: built-in trigger lookup is pinned case-sensitive, and `builtin:nonexistent` is pinned to skip without DB access.

## Quality gates

- 61 new tests, all green; full command-adjacent suite (795 web tests + 8 db tests) green
- Baseline was green pre-change (271 command tests); DB fully mocked; route/pure-module level only (no React render tests, per worktree constraint)
- `bun run typecheck` clean (13/13), `bun run lint` clean (13/13), no `any`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Documentation**
  * Added launch checklist, QA procedures, and rollout/flip plan for the Universal Commands feature.

* **Tests**
  * Added extensive edge-case tests covering command resolution, command-chip/message parsing, agent-mention interactions, lifecycle degradations (deleted/trashed/missing/disabled/restricted), hostile/forged inputs, schema cascade behavior, and resolver robustness to ensure safe degradation and no sensitive leaks.

## Review follow-ups (post-open)

- **Codex P2 ×2 (both resolved):** the dogfooding drive-count SQL now uses NULL-excluding `count(DISTINCT drive_id)`, and the cross-token isolation test asserts the agent lookup's WHERE filters by exactly the mention id (structured operator mocks).
- **Self-review pass (7-angle finder sweep):** strengthened the resolver chip+mention test to assert mention ids never reach the command lookup; extracted a shared `armDefaultMocks()` helper in the responder edge-case suite; doc now spells out full paths for Phase 5-owned files and records the resolve route's per-id permission loop (N+1, capped at `MAX_IDS=50`) as a post-launch batching follow-up.

### CODE REVIEW

File: docs/specs/universal-commands-launch-checklist.md [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+transcripts, `/api/commands/resolve` for chip state, execution pills on AI
+responses, and command execution itself.
+
+### 2.4 Dogfooding exit criteria (all must hold before flipping)
+
+Concrete and checkable; run the checks over a trailing **7-day** window.
+
+1. **Creation volume:** ≥ 10 non-builtin commands exist across ≥ 3 drives
+   and ≥ 4 distinct authors.
+   `SELECT count(*), count(DISTINCT coalesce(drive_id, 'personal')), count(DISTINCT created_by_id) FROM commands;`
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply docs/specs/universal-commands-launch-checklist.md]: Fixed in 622644dcc — the criterion now uses `count(DISTINCT drive_id)` (which excludes NULLs in Postgres), so personal commands can no longer inflate the drive count, and the criterion text now says "drive commands span ≥ 3 distinct drives" explicitly.

File: apps/web/src/lib/channels/__tests__/agent-mention-responder-edge-cases.test.ts [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+  it('never treats the mentioned agent id as a command or the command id as a mention', async () => {
+    await triggerMentionedAgentResponses(params);
+    // Command resolution queried exactly the chip's commandId path once.
+    expect(mockCommandsFindFirst).toHaveBeenCalledTimes(1);
+    // The agent lookup received the mention id, not the command id.
+    const agentLookupCall = mockPagesFindMany.mock.calls.find(
+      (call) => (call[0] as { columns?: Record<string, boolean> })?.columns?.enabledTools
+    );
+    expect(agentLookupCall).toBeDefined();
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply apps/web/src/lib/channels/__tests__/agent-mention-responder-edge-cases.test.ts:212]: Fixed in 622644dcc — the operator mocks now return structured objects (`{op, field, values}`), and the test asserts the agent lookup's WHERE contains `inArray` with exactly `[AGENT_ID]`. A regression that feeds the command id into mention resolution now fails this assertion rather than passing on column shape alone.
